src/ModelContextProtocol/McpChatClientBuilderExtensions.cs

```diff
@@ -0,0 +1,230 @@
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                // Dispose of the HTTP client if it was created by this client.
+                if (_ownsHttpClient)
+                {
+                    _httpClient?.Dispose();
+                }
+
+                if (_mcpClientTasks is not null)
+                {
+                    // Dispose of all cached MCP clients.
+                    foreach (var clientTask in _mcpClientTasks.Values)
+                    {
+#if NETSTANDARD2_0
+                        if (clientTask.Status == TaskStatus.RanToCompletion)
+#else
+                        if (clientTask.IsCompletedSuccessfully)
+#endif
+                        {
+                            _ = clientTask.Result.DisposeAsync();
+                        }
+                    }
+
+                    _mcpClientTasks.Clear();
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        private Task<McpClient> CreateMcpClientAsync(Uri serverAddress, string serverName, string? authorizationToken)
+        {
+            if (_mcpClientTasks is null)
+            {
+                _mcpClientTasks = new ConcurrentDictionary<string, Task<McpClient>>(StringComparer.OrdinalIgnoreCase);
+            }
+
+            // Note: We don't pass cancellationToken to the factory because the cached task should not be tied to any single caller's cancellation token.
+            // Instead, callers can cancel waiting for the task, but the connection attempt itself will complete independently.
+            return _mcpClientTasks.GetOrAdd(serverAddress.ToString(), _ => CreateMcpClientCoreAsync(serverAddress, serverName, authorizationToken, CancellationToken.None));
+        }
+
+        private async Task<McpClient> CreateMcpClientCoreAsync(Uri serverAddress, string serverName, string? authorizationToken, CancellationToken cancellationToken)
+        {
+            var serverAddressKey = serverAddress.ToString();
+            try
+            {
+                var transport = new HttpClientTransport(new HttpClientTransportOptions
+                {
+                    Endpoint = serverAddress,
+                    Name = serverName,
+                    AdditionalHeaders = authorizationToken is not null
+                        // Update to pass all headers once https://github.com/dotnet/extensions/pull/7053 is available.
+                        ? new Dictionary<string, string>() { { "Authorization", $"Bearer {authorizationToken}" } }
+                        : null,
+                }, _httpClient, _loggerFactory);
+
+                return await McpClient.CreateAsync(transport, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+            catch
+            {
+                // Remove the failed task from cache so subsequent requests can retry
+                _mcpClientTasks?.TryRemove(serverAddressKey, out _);
```

Comment:
It may not have been added to the cache yet, if all the awaits until this point are for things that completed synchronously, or given other race conditions.